src/datasets/arrow_dataset.py

```diff
@@ -1577,19 +1578,33 @@ def apply_function_on_filtered_inputs(inputs, indices, check_same_num_examples=F
                             "Using `.map` in batched mode on a dataset with attached indexes is allowed only if it doesn't create or remove existing examples. You can first run `.drop_index() to remove your index and then re-add it."
                         )
                     if update_data:
+                        if i == 0:
+                            cache_file_name = determine_cache_file_name(cache_file_name)
+                            is_cached = check_if_cached(cache_file_name)
+                            if is_cached:
+                                break
+                            else:
+                                buf_writer, writer, tmp_file = init_buffer_and_writer(cache_file_name)
                         batch = cast_to_python_objects(batch)
                         writer.write_batch(batch)
-            if update_data:
+            if update_data and writer is not None:
                 writer.finalize()  # close_stream=bool(buf_writer is None))  # We only close if we are writing in a file
         except (Exception, KeyboardInterrupt):
             if update_data:
-                writer.finalize()
-            if update_data and tmp_file is not None:
-                tmp_file.close()
-                if os.path.exists(tmp_file.name):
-                    os.remove(tmp_file.name)
+                if writer is not None:
+                    writer.finalize()
+                if tmp_file is not None:
+                    tmp_file.close()
+                    if os.path.exists(tmp_file.name):
+                        os.remove(tmp_file.name)
             raise
 
+        if update_data and is_cached:
+            logger.warning("Loading cached processed dataset at %s", cache_file_name)
+            info = self.info.copy()
+            info.features = features
+            return Dataset.from_file(cache_file_name, info=info, split=self.split)
```

Comment:
If it's ok to have a return statement inside the for loop in terms of style, then we can put this code inside the loop and remove the `is_cached` variable (for better readability?)

Reply:
Sure looks good !